crates/turbo-tasks-memory/src/task.rs

```diff
@@ -92,15 +106,160 @@ pub struct Task {
     /// The type of the task
     ty: TaskType,
     /// The mutable state of the task
-    state: RwLock<TaskState>,
+    /// Unset state is equal to a Dirty task that has not been executed yet
+    state: RwLock<TaskMetaState>,
+}
+
+enum TaskMetaState {
+    Full(Box<TaskState>),
+    Partial(Box<PartialTaskState>),
+    Unloaded(UnloadedTaskState),
+}
+
+impl Default for TaskMetaState {
+    fn default() -> Self {
+        Self::Unloaded(UnloadedTaskState::default())
+    }
+}
+
+impl TaskMetaState {
+    fn into_unwrap_partial(self) -> PartialTaskState {
```

Comment:
Nit: Since these return an owned type, I would just call them `unwrap_*`, e.g. [Result::unwrap_err](https://doc.rust-lang.org/std/result/enum.Result.html#method.unwrap_err)